README.md
![Tracing — Structured, application-level diagnostics][splash]

[splash]: https://raw.githubusercontent.com/tokio-rs/tracing/master/assets/splash.svg

# tracing-appender

Writers for logging events and spans

[![Crates.io][crates-badge]][crates-url]
[![Documentation][docs-badge]][docs-url]
[![Documentation (master)][docs-master-badge]][docs-master-url]
[![MIT licensed][mit-badge]][mit-url]
[![Build Status][actions-badge]][actions-url]
[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-url] | [Chat][discord-url]

[crates-badge]: https://img.shields.io/crates/v/tracing-appender.svg

[crates-url]: https://crates.io/crates/tracing-appender/0.2.2

[docs-badge]: https://docs.rs/tracing-appender/badge.svg

[docs-url]: https://docs.rs/tracing-appender/0.2.2

[docs-master-badge]: https://img.shields.io/badge/docs-master-blue

[docs-master-url]: https://tracing.rs/tracing-appender

[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg

[mit-url]: ./LICENSE

[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg

[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI

[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white

[discord-url]: https://discord.gg/EeF3cQw

## Overview

[`tracing`][tracing] is a framework for instrumenting Rust programs to
collect structured, event-based diagnostic information. `tracing-appender`
allows events and spans to be recorded in a non-blocking manner through a
dedicated logging thread. It also provides a [`RollingFileAppender`][file_appender]
that can be used with _or_ without the non-blocking writer.

*Compiler support: [requires `rustc` 1.53+][msrv]*

[msrv]: #supported-rust-versions

## Usage

Add the following to your `Cargo.toml`:

```toml
tracing-appender = "0.2"
```

This crate can be used in a few ways to record spans/events:

- Using a [`RollingFileAppender`][file_appender] to write to a log file.
  This is a blocking operation.
- Using *any* type implementing [`std::io::Write`][write] in a
  non-blocking fashion.
- Using [`NonBlocking`][non_blocking] and [`RollingFileAppender`][file_appender]
  together to write to log files in a non-blocking fashion.

## Rolling File Appender

```rust
fn main(){
    let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
}
```

This creates an hourly rotating file appender that writes to
`/some/directory/prefix.log.YYYY-MM-DD-HH`. [`Rotation::DAILY`] and
[`Rotation::NEVER`] are the other available options.

The file appender implements [`std::io::Write`][write]. To be used with
[`tracing_subscriber::FmtSubscriber`][fmt_subscriber], it must be combined
with a [`MakeWriter`][make_writer] implementation to be able to record
tracing spans/event.

The [rolling] module's documentation provides more detail on how to use
this file appender.

## Non-Blocking Writer

The example below demonstrates the construction of a `non_blocking` writer
with an implementation of [`std::io::Writer`][write].

```rust
use std::io::Error;

struct TestWriter;

impl std::io::Write for TestWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let buf_len = buf.len();
    
        println!("{:?}", buf);
        Ok(buf_len)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

fn main() {
    let (non_blocking, _guard) = tracing_appender::non_blocking(TestWriter);
    tracing_subscriber::fmt().with_writer(non_blocking).init();
}
```

**Note:** `_guard` is a [`WorkerGuard`][guard] which is returned by
`tracing_appender::non_blocking` to ensure buffered logs are flushed to
their output in the case of abrupt terminations of a process. See
[`WorkerGuard`][guard] module for more details.

The example below demonstrates the construction of a
[`tracing_appender::non_blocking`][non_blocking] writer constructed with
a [`std::io::Write`][write]:

```rust
fn main() {
    let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
    tracing_subscriber::fmt()
        .with_writer(non_blocking)
        .init();
}
```

The [non_blocking] module's documentation provides more detail on how to
use `non_blocking`.

## Non-Blocking Rolling File Appender

```rust
fn main() {
    let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
    let (non_blocking, _guard) = tracing_appender::non_blocking(file_appender);
   tracing_subscriber::fmt()
       .with_writer(non_blocking)
       .init();
}
```

[tracing]: https://docs.rs/tracing/latest/tracing/

[make_writer]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html

[write]: https://doc.rust-lang.org/std/io/trait.Write.html

[non_blocking]: https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/index.html

[rolling]: https://docs.rs/tracing-appender/latest/tracing_appender/rolling/index.html

[guard]: https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/struct.WorkerGuard.html

[file_appender]: https://docs.rs/tracing-appender/latest/tracing_appender/rolling/struct.RollingFileAppender.html

[fmt_subscriber]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Subscriber.html

## Supported Rust Versions

`tracing-appender` is built against the latest stable release. The minimum supported
version is 1.53. The current `tracing-appender` version is not guaranteed to build on
Rust versions earlier than the minimum supported version.

Tracing follows the same compiler support policies as the rest of the Tokio
project. The current stable Rust compiler and the three most recent minor
versions before it will always be supported. For example, if the current
stable compiler version is 1.45, the minimum supported version will not be
increased past 1.42, three minor versions prior. Increasing the minimum
supported compiler version is not considered a semver breaking change as
long as doing so complies with this policy.

## License

This project is licensed under the [MIT license](./LICENSE).

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in Tokio by you, shall be licensed as MIT, without any additional
terms or conditions.


CHANGELOG.md
# 0.2.2 (March 17, 2022)

This release fixes a bug in `RollingFileAppender` that could result
in a failure to rotate the log file, or in panics in debug mode.

### Fixed

- **rolling**: Fixed a panic that prohibited rolling files over. ([#1989])

[#1989]: https://github.com/tokio-rs/tracing/pull/1989

# 0.2.1 (February 28, 2022)

This release adds an implementation of the `MakeWriter` trait for
`RollingFileAppender`, allowing it to be used without wrapping in a
`NonBlocking` writer.

This release increases the minimum supported Rust version to 1.53.0.

### Added

- **rolling**: Added `MakeWriter` implementation for `RollingFileAppender`
  ([#1760])

### Changed

- Updated minimum supported Rust version (MSRV) to 1.53.0 ([#1851])
- `parking_lot`: updated to v0.12 ([#1878])

### Fixed

- Fixed several documentation typos and issues ([#1780], [d868054], [#1943])

[#1760]: https://github.com/tokio-rs/tracing/pull/1760

[#1851]: https://github.com/tokio-rs/tracing/pull/1851

[#1878]: https://github.com/tokio-rs/tracing/pull/1878

[#1943]: https://github.com/tokio-rs/tracing/pull/1943

[d868054]: https://github.com/tokio-rs/tracing/commit/d8680547b509978c7113c8f7e19e9b00c789c698

# 0.2.0 (October 22, 2021)

This breaking change release adds support for the new v0.3.x series of
`tracing-subscriber`. In addition, it resolves the security advisory for the
`chrono` crate, [RUSTSEC-2020-0159].

This release increases the minimum supported Rust version to 1.51.0.

### Breaking Changes

- Updated `tracing-subscriber` to v0.3.x ([#1677])
- Changed `NonBlocking::error_counter` to return an `ErrorCounter` type, rather
  than an `Arc<AtomicU64>` ([#1675])

### Changed

- Updated `tracing-subscriber` to v0.3.x ([#1677])

### Fixed

- **non-blocking**: Fixed compilation on 32-bit targets ([#1675])
- **rolling**: Replaced `chrono` dependency with `time` to resolve
  [RUSTSEC-2020-0159] ([#1652])
- **rolling**: Fixed an issue where `RollingFileAppender` would fail to print
  errors that occurred while flushing a previous logfile ([#1604])

Thanks to new contributors @dzvon and @zvkemp for contributing to this release!

[RUSTSEC-2020-0159]: https://rustsec.org/advisories/RUSTSEC-2020-0159.html

[#1677]: https://github.com/tokio-rs/tracing/pull/1677

[#1675]: https://github.com/tokio-rs/tracing/pull/1675

[#1652]: https://github.com/tokio-rs/tracing/pull/1675

[#1604]: https://github.com/tokio-rs/tracing/pull/1604

# 0.1.2 (December 28, 2020)

### Changed

- **non_blocking**: Updated `crossbeam-channel` dependency to 0.5 (#1031)

### Fixed

- **non_blocking**: Fixed a race condition when logging on shutdown (#1125)
- Several documentation improvements (#1109, #1110, #941, #953)

# 0.1.1 (July 20, 2020)

### Added

- **rolling**: `minutely` rotation schedule to rotate the log file once every
  minute (#748)

### Fixed

- Fixed broken links in docs (#718)
- `tracing-appender` now only enables the necessary `tracing-subscriber`'s
  feature flags, rather than all of them (#779)

Thanks to new contributors @ericjheinz and @sourcefrog for contributing
to this release!

# 0.1.0 (May 5, 2020)

- Initial release

src/worker.rs
use crate::Msg;
use crossbeam_channel::{Receiver, RecvError, TryRecvError};
use std::fmt::Debug;
use std::io::Write;
use std::{io, thread};

pub(crate) struct Worker<T: Write + Send + Sync + 'static> {
    writer: T,
    receiver: Receiver<Msg>,
    shutdown: Receiver<()>,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub(crate) enum WorkerState {
    Empty,
    Disconnected,
    Continue,
    Shutdown,
}

impl<T: Write + Send + Sync + 'static> Worker<T> {
    pub(crate) fn new(receiver: Receiver<Msg>, writer: T, shutdown: Receiver<()>) -> Worker<T> {
        Self {
            writer,
            receiver,
            shutdown,
        }
    }

    fn handle_recv(&mut self, result: &Result<Msg, RecvError>) -> io::Result<WorkerState> {
        match result {
            Ok(Msg::Line(msg)) => {
                self.writer.write_all(msg)?;
                Ok(WorkerState::Continue)
            }
            Ok(Msg::Shutdown) => Ok(WorkerState::Shutdown),
            Err(_) => Ok(WorkerState::Disconnected),
        }
    }

    fn handle_try_recv(&mut self, result: &Result<Msg, TryRecvError>) -> io::Result<WorkerState> {
        match result {
            Ok(Msg::Line(msg)) => {
                self.writer.write_all(msg)?;
                Ok(WorkerState::Continue)
            }
            Ok(Msg::Shutdown) => Ok(WorkerState::Shutdown),
            Err(TryRecvError::Empty) => Ok(WorkerState::Empty),
            Err(TryRecvError::Disconnected) => Ok(WorkerState::Disconnected),
        }
    }

    /// Blocks on the first recv of each batch of logs, unless the
    /// channel is disconnected. Afterwards, grabs as many logs as
    /// it can off the channel, buffers them and attempts a flush.
    pub(crate) fn work(&mut self) -> io::Result<WorkerState> {
        // Worker thread yields here if receive buffer is empty
        let mut worker_state = self.handle_recv(&self.receiver.recv())?;

        while worker_state == WorkerState::Continue {
            let try_recv_result = self.receiver.try_recv();
            let handle_result = self.handle_try_recv(&try_recv_result);
            worker_state = handle_result?;
        }
        self.writer.flush()?;
        Ok(worker_state)
    }

    /// Creates a worker thread that processes a channel until it's disconnected
    pub(crate) fn worker_thread(mut self) -> thread::JoinHandle<()> {
        thread::spawn(move || {
            loop {
                match self.work() {
                    Ok(WorkerState::Continue) | Ok(WorkerState::Empty) => {}
                    Ok(WorkerState::Shutdown) | Ok(WorkerState::Disconnected) => {
                        let _ = self.shutdown.recv();
                        break;
                    }
                    Err(_) => {
                        // TODO: Expose a metric for IO Errors, or print to stderr
                    }
                }
            }
            if let Err(e) = self.writer.flush() {
                eprintln!("Failed to flush. Error: {}", e);
            }
        })
    }
}


src/lib.rs
//! Writers for logging events and spans
//!
//! # Overview
//!
//! [`tracing`](https://docs.rs/tracing) is a framework for structured,
//! event-based diagnostic information.
//! `tracing-appender` allows events and spans to be recorded in a non-blocking manner through
//! a dedicated logging thread. It also provides a [RollingFileAppender] that can
//! be used with _or_ without the non-blocking writer.
//!
//! *Compiler support: [requires `rustc` 1.53+][msrv]*
//!
//! [msrv]: #supported-rust-versions
//! [rolling]: rolling
//! [RollingFileAppender]: rolling::RollingFileAppender
//! [tracing_subscriber::FmtSubscriber]: tracing_subscriber::FmtSubscriber
//! [Write]: Write
//! [NonBlocking]: NonBlocking
//! [WorkerGuard]: WorkerGuard
//! [MakeWriter]: tracing_subscriber::fmt::MakeWriter
//! [non_blocking]: fn@non_blocking
//! [non_blocking module]: mod@non_blocking
//! [daily]: rolling::daily
//! [never]: rolling::never
//!
//!
//! # Usage
//!
//! Add the following to your `Cargo.toml`:
//! ```toml
//! tracing-appender = "0.2"
//! ```
//!
//! This crate can be used in a few ways to record spans/events:
//!  - Using a [RollingFileAppender] to perform writes to a log file. This will block on writes.
//!  - Using *any* type implementing [Write] in a non-blocking fashion.
//!  - Using a combination of [NonBlocking] and [RollingFileAppender] to allow writes to a log file
//! without blocking.
//!
//! ## Rolling File Appender
//!
//! ```rust
//! # fn docs() {
//! let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
//! # }
//! ```
//! This creates an hourly rotating file appender that writes to
//! `/some/directory/prefix.YYYY-MM-DD-HH.log`.
//! [daily]  and [never] are the other available options.
//!
//! The file appender implements [Write]. To be used with [tracing_subscriber::FmtSubscriber],
//! it must be combined with a [MakeWriter] implementation to be able to record tracing spans/event.
//!
//! The [rolling] module's documentation provides more detail on how to use this file appender.
//!
//! ## Non-Blocking Writer
//!
//! The example below demonstrates the construction of a `non_blocking` writer
//! with `std::io::stdout()`,
//! which implements [MakeWriter].
//!
//! ```rust
//! # fn doc() {
//! let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
//! tracing_subscriber::fmt()
//!     .with_writer(non_blocking)
//!     .init();
//! # }
//! ```
//! **Note:** `_guard` is a [WorkerGuard]
//! which is returned by [non_blocking]
//! to ensure buffered logs are flushed to their output
//! in the case of abrupt terminations of a process.
//! See [WorkerGuard] for more details.
//!
//! The example below demonstrates the construction of a [non_blocking]
//! writer constructed with a [Write]:
//!
//! ```rust
//! use std::io::Error;
//!
//! struct TestWriter;
//!
//! impl std::io::Write for TestWriter {
//!     fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
//!         let buf_len = buf.len();
//!         println!("{:?}", buf);
//!         Ok(buf_len)
//!     }
//!
//!     fn flush(&mut self) -> std::io::Result<()> {
//!         Ok(())
//!     }
//! }
//!
//! # fn doc() {
//! let (non_blocking, _guard) = tracing_appender::non_blocking(TestWriter);
//! tracing_subscriber::fmt()
//!     .with_writer(non_blocking)
//!     .init();
//! # }
//! ```
//!
//! The [non_blocking module]'s documentation provides
//! more detail on how to use `non_blocking`.
//!
//!
//! ## Non-Blocking Rolling File Appender
//!
//! ```rust
//! # fn docs() {
//! let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
//! let (non_blocking, _guard) = tracing_appender::non_blocking(file_appender);
//! tracing_subscriber::fmt()
//!     .with_writer(non_blocking)
//!     .init();
//! # }
//! ```
//!
//! ## Supported Rust Versions
//!
//! `tracing-appender` is built against the latest stable release. The minimum supported
//! version is 1.53. The current `tracing-appender` version is not guaranteed to build on
//! Rust versions earlier than the minimum supported version.
//!
//! Tracing follows the same compiler support policies as the rest of the Tokio
//! project. The current stable Rust compiler and the three most recent minor
//! versions before it will always be supported. For example, if the current
//! stable compiler version is 1.45, the minimum supported version will not be
//! increased past 1.42, three minor versions prior. Increasing the minimum
//! supported compiler version is not considered a semver breaking change as
//! long as doing so complies with this policy.
//!
#![doc(html_root_url = "https://docs.rs/tracing-appender/0.2.2")]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/tokio-rs/tracing/master/assets/logo-type.png"
)]
#![doc(issue_tracker_base_url = "https://github.com/tokio-rs/tracing/issues/")]
#![cfg_attr(docsrs, deny(rustdoc::broken_intra_doc_links))]
#![warn(
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms,
    unreachable_pub,
    bad_style,
    dead_code,
    improper_ctypes,
    non_shorthand_field_patterns,
    no_mangle_generic_items,
    overflowing_literals,
    path_statements,
    patterns_in_fns_without_body,
    private_in_public,
    unconditional_recursion,
    unused,
    unused_allocation,
    unused_comparisons,
    unused_parens,
    while_true
)]

use crate::non_blocking::{NonBlocking, WorkerGuard};

use std::io::Write;

pub mod non_blocking;

pub mod rolling;

mod worker;

pub(crate) mod sync;

/// Convenience function for creating a non-blocking, off-thread writer.
///
/// See the [`non_blocking` module's docs][non_blocking]'s for more details.
///
/// [non_blocking]: ./non_blocking/index.html
///
/// # Examples
///
/// ``` rust
/// # fn docs() {
/// let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
/// let subscriber = tracing_subscriber::fmt().with_writer(non_blocking);
/// tracing::subscriber::with_default(subscriber.finish(), || {
///    tracing::event!(tracing::Level::INFO, "Hello");
/// });
/// # }
/// ```
pub fn non_blocking<T: Write + Send + Sync + 'static>(writer: T) -> (NonBlocking, WorkerGuard) {
    NonBlocking::new(writer)
}

#[derive(Debug)]
pub(crate) enum Msg {
    Line(Vec<u8>),
    Shutdown,
}
